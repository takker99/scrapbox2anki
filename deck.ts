import type { Result } from "./deps/scrapbox.ts";
import { parse } from "./deps/csv.ts";
import type { Deck } from "./deps/deno-anki.ts";

/** deckデータの書式が不正だったときに投げるエラー */
export interface InvalidDeckError {
  name: "InvalidDeckError";
  message: string;
}

export const parseDeck = (csv: string): Result<Deck, InvalidDeckError> => {
  const data = parse(csv);
  const csvData = new Map<string, string>(data.map(([f, s]) => [f, s]));
  const name = csvData.get("name");
  if (!name) return { ok: false, value: makeError("Deck name not found.") };
  const id_ = csvData.get("id");
  if (!id_) return { ok: false, value: makeError("Deck id not found.") };
  const id = parseInt(id_);
  if (isNaN(id)) {
    return { ok: false, value: makeError("Deck id is not number.") };
  }

  const deck: Deck = { name, id };
  const description = csvData.get("description");
  if (description) deck.description = description;

  return { ok: true, value: deck };
};

const makeError = (message: string): InvalidDeckError => ({
  name: "InvalidDeckError",
  message,
});
