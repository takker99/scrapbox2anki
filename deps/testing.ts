export * from "https://deno.land/std@0.186.0/testing/asserts.ts";
export * from "https://deno.land/std@0.186.0/testing/snapshot.ts";
