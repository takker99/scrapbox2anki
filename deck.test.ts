/// <reference lib="deno.ns" />

import { assertSnapshot } from "./deps/testing.ts";
import { parseDeck } from "./deck.ts";
import lines from "./sample-deck1.json" assert { type: "json" };
import lines2 from "./sample-page1.json" assert { type: "json" };

Deno.test("parseDeck()", async (t) => {
  await t.step("not found", async (t) => {
    await assertSnapshot(t, parseDeck(lines));
  });
  await t.step("found", async (t) => {
    await assertSnapshot(t, parseDeck(lines2));
  });
});
