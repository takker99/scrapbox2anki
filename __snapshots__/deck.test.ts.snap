export const snapshot = {};

snapshot[`parseDeck() 1`] = `
{
  ok: true,
  value: {
    description: "システム英単語系の問題集",
    id: 44845749858349580,
    name: "system-english",
    updated: 1676275415,
  },
}
`;
